src/groupeddataframe/grouping.jl

```diff
@@ -181,6 +197,242 @@ end
 Base.getindex(gd::GroupedDataFrame, idxs::Colon) =
     GroupedDataFrame(gd.parent, gd.cols, gd.groups, gd.idx, gd.starts, gd.ends)
 
+
+# Eltypes of the grouping columns
+_grouptypes(gd::GroupedDataFrame) = Tuple(eltype(gd.parent[!, c]) for c in gd.cols)
+
+# Iterator over values of grouping columns for each group
+function _groupvalues(gd::GroupedDataFrame)
+    T = Tuple{_grouptypes(gd)...}
+    return (T(gd.parent[gd.idx[s], gd.cols]) for s in gd.starts)
+end
+
+# Get values of grouping columns for single group by its index
+_groupvalues(gd::GroupedDataFrame, i::Integer) = Tuple(gd.parent[gd.idx[gd.starts[i]], gd.cols])
+
+# Get value of grouping column by group and column index
+function _groupvalue(gd::GroupedDataFrame, groupidx::Integer, colidx::Integer)
+    return gd.parent[gd.idx[gd.starts[groupidx]], gd.cols[colidx]]
+end
+
+
+"""
+    GroupKey
+
+Key for one of the groups of a [`GroupedDataFrame`](@ref). Contains the values
+of the corresponding grouping columns and behaves similarly to a `NamedTuple`,
+but using it to index its `GroupedDataFrame` is much more effecient than using the
+equivalent `Tuple` or `NamedTuple`.
+
+Instances of this type are returned by `keys(::GroupedDataFrame)` and are not
+meant to be constructed directly.
+
+See [`keys(::GroupedDataFrame)`](@ref) for more information.
+"""
+struct GroupKey{N, T<:Tuple}
+    parent::GroupedDataFrame
```

Comment:
`GroupedDataFrame` is a parametric type so it should be also moved as parameter to `GroupKey`.